src/pseudo/PspUpf.jl

```diff
@@ -65,164 +65,169 @@ Does not support:
-                                  identifier, description)
+    r2_ρion_interp = linear_interpolation((rgrid,), r2_ρion)
+    r2_ρcore_interp = linear_interpolation((rgrid,), r2_ρcore)
+
+    PspUpf{eltype(rgrid),typeof(vloc_interp)}(Zion, lmax, rgrid, drgrid, vloc,
+        r2_projs, h, r2_pswfcs, pswfc_occs, r2_ρion, r2_ρcore, vloc_interp,
+        r2_projs_interp, r2_ρion_interp, r2_ρcore_interp, rcut, ircut,
+        identifier, description
+    )
 end
 
 charge_ionic(psp::PspUpf) = psp.Zion
-has_valence_density(psp::PspUpf) = !all(iszero, psp.r2_4π_ρion)
-has_core_density(psp::PspUpf) = !all(iszero, psp.ρcore)
+has_valence_density(psp::PspUpf) = !all(iszero, psp.r2_ρion)
+has_core_density(psp::PspUpf) = !all(iszero, psp.r2_ρcore)
 
-# Note: UPFs store `r[j] β_{li}(r[j])`, so r=0 is undefined and will error.
-function eval_psp_projector_real(psp::PspUpf, i, l, r::T)::T where {T <: Real}
-    psp.r_projs_interp[l+1][i](r) / r
+function eval_psp_projector_real(psp::PspUpf, i, l, r::T)::T where {T<:Real}
+    psp.r2_projs_interp[l+1][i](r) / r^2
 end
 
-# For UPFs, the integral is transformed to the following sum:
-# 4π Σ{k} j_l(q r[k]) (r[k]^2 p_{il}(r[k]) dr[k])
-function eval_psp_projector_fourier(psp::PspUpf, i, l, q::T)::T where {T <: Real}
-    r2_proj_dr = psp.r2_projs_dr[l+1][i]
-    s = zero(T)
-    @inbounds for ir = eachindex(r2_proj_dr)
-        s += sphericalbesselj_fast(l, q * psp.rgrid[ir]) * r2_proj_dr[ir]
-    end
-    4T(π) * s
+function eval_psp_projector_fourier(psp::PspUpf, i, l, q::T)::T where {T<:Real}
+    # The projectors may have been cut off before the end of the radial mesh
+    # by PseudoPotentialIO because UPFs list a radial cutoff index for these
+    # functions after which they are strictly zero in the file.
+    ircut = min(psp.ircut, length(psp.r2_projs[l+1][i]))
+    rgrid = @view psp.rgrid[1:ircut]
+    r2_proj = psp.r2_projs[l+1][i][1:ircut]
+    return hankel(rgrid, r2_proj, l, q)
 end
 
-eval_psp_local_real(psp::PspUpf, r::T) where {T <: Real} = psp.vloc_interp(r)
+function eval_psp_pswfc_real(psp::PspUpf, i, l, r::T)::T where {T<:Real}
+    psp.r2_pswfcs_interp[l+1][i](r) / r^2
+end
 
-# For UPFs, the integral is transformed to the following sum:
-# 4π/q (Σ{i} sin(q r[i]) (r[i] V(r[i]) + Z) dr[i] - Z/q)
-function eval_psp_local_fourier(psp::PspUpf, q::T)::T where {T <: Real}
-    s = zero(T)
-    @inbounds for ir = eachindex(psp.r_vloc_corr_dr)
-        s += sin(q * psp.rgrid[ir]) * psp.r_vloc_corr_dr[ir]
-    end
-    4T(π) * (s - psp.Zion / q) / q
+function eval_psp_pswfc_fourier(psp::PspUpf, i, l, q::T)::T where {T<:Real}
+    # Pseudo-atomic wavefunctions are _not_ currently cut off like the other
+    # quantities. They are the reason that PseudoDojo UPF files have a much
+    # larger radial grid than their psp8 counterparts.
+    # If issues arise, try cutting them off too.
+    ircut = length(psp.r2_pswfcs[l+1][i])
+    rgrid = @view psp.rgrid[1:ircut]
```

Comment:
maybe then don't use the variable name `ircut` to avoid someone missing this point when glossing over the code.